File: ios/Classes/lib/DartBackgroundExecutor.swift
//
//  DartBackgroundService.swift
//  awesome_notifications
//
//  Created by Rafael Setragni on 04/06/21.
//
#if !ACTION_EXTENSION
import UIKit
import Foundation
import Flutter
import IosAwnCore

public class DartBackgroundExecutor: BackgroundExecutor {
    private let TAG = "DartBackgroundExecutor"
    
    public let silentDataQueue:SynchronizedArray = SynchronizedArray<SilentActionRequest>()
        
    private var backgroundEngine: FlutterEngine?
    private var backgroundChannel: FlutterMethodChannel?
    
    static var registrar: FlutterPluginRegistrar?
    
    private var _isRunning = false
    public var isRunning:Bool {
        get { return _isRunning }
    }
    public var isNotRunning: Bool {
        get { return !_isRunning }
    }
    
    private var actionReceived:ActionReceived?
    
    // ************** Awesome Extensions ***********************00
    
    public required init(){}
    
    public static func extendCapabilities(usingFlutterRegistrar registrar:FlutterPluginRegistrar){
        self.registrar = registrar
        AwesomeNotifications.backgroundClassType = DartBackgroundExecutor.self
    }
    
    // ************** IOS EVENTS LISTENERS ************************
        
    var dartCallbackHandle:Int64 = 0
    var silentCallbackHandle:Int64 = 0
    
    public func runBackgroundProcess(
        silentActionRequest:SilentActionRequest,
        dartCallbackHandle:Int64,
        silentCallbackHandle:Int64
    ){
        addSilentActionRequest(silentActionRequest)
       
        if(!self._isRunning){
            self._isRunning = true
            
            self.silentCallbackHandle = silentCallbackHandle
            self.dartCallbackHandle = dartCallbackHandle
            
            runBackgroundThread(
                silentCallbackHandle: silentCallbackHandle,
                dartCallbackHandle: dartCallbackHandle)
        }
    }
    
    public func addSilentActionRequest(_ silentActionRequest:SilentActionRequest){
        silentDataQueue.append(silentActionRequest)
    }
    
    public func onMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult){
        
        switch(call.method){
            
        case Definitions.CHANNEL_METHOD_PUSH_NEXT:
                dischargeNextSilentExecution()
                result(true)
                break
            
            default:
                result(
                    FlutterError.init(
                        code: TAG,
                        message: "\(call.method) not implemented",
                        details: call.method
                    )
                )
                result(false)
                break
        }
    }
    
    func runBackgroundThread(
        silentCallbackHandle:Int64,
        dartCallbackHandle:Int64
    ) {
        
        guard let dartCallbackInfo = FlutterCallbackCache.lookupCallbackInformation(dartCallbackHandle) else {
            Logger.e(TAG, "There is no valid callback info to handle dart channels.")
            closeBackgroundIsolate()
            return
        }
        
        guard FlutterCallbackCache.lookupCallbackInformation(silentCallbackHandle) != nil else {
            Logger.e(TAG, "There is no valid callback info to handle silent data.")
            closeBackgroundIsolate()
            return
        }
        
        if(self.backgroundEngine != nil){
            Logger.d(TAG, "Background isolate already started.")
            closeBackgroundIsolate()
            return
        }
        
        // Flutter engine still doesn't support running in background thread
        if Thread.isMainThread {
            createNewFlutterEngine(dartCallbackInfo: dartCallbackInfo)
        }
        else {
            DispatchQueue.main.async(execute: {
                self.createNewFlutterEngine(dartCallbackInfo: dartCallbackInfo)
            })
        }
    }
    
    func createNewFlutterEngine(
        dartCallbackInfo:FlutterCallbackInformation
    ){
        do {
            
            self.backgroundEngine = FlutterEngine(
                name: "AwesomeNotificationsBgEngine",
                project: nil,
                allowHeadlessExecution: true
            )
            
            if self.backgroundEngine == nil {
                Logger.e(self.TAG, "Flutter background engine is not available.")
                self.closeBackgroundIsolate()
            }
            else {
                self.backgroundEngine!.run(
                    withEntrypoint: dartCallbackInfo.callbackName,
                    libraryURI: dartCallbackInfo.callbackLibraryPath)
                
                SwiftAwesomeNotificationsPlugin
                    .flutterRegistrantCallback?(self.backgroundEngine!)
                
                DartAwesomeNotificationsExtension.setRegistrar(flutterEngine: backgroundEngine)
                DartAwesomeNotificationsExtension.initialize()
                try AwesomeNotifications.loadExtensions()
                
                self.initializeReverseMethodChannel(
                    backgroundEngine: self.backgroundEngine!)
            }
            
        } catch {
            Logger.e(TAG, error.localizedDescription)
        }
    }
    
    func initializeReverseMethodChannel(backgroundEngine: FlutterEngine){
        
        self.backgroundChannel = FlutterMethodChannel(
            name: Definitions.DART_REVERSE_CHANNEL,
            binaryMessenger: backgroundEngine.binaryMessenger
        )
        
        self.backgroundChannel!.setMethodCallHandler(onMethodCall)
    }
    
    func closeBackgroundIsolate() {
        _isRunning = false
        
        self.backgroundEngine?.destroyContext()
        self.backgroundEngine = nil
        
        self.backgroundChannel = nil
        Logger.i(TAG, "FlutterEngine instance terminated.")
    }
    
    func dischargeNextSilentExecution(){
        if let silentDataRequest:SilentActionRequest = silentDataQueue.first {
            silentDataQueue.remove(at: 0)
            self.executeDartCallbackInBackgroundIsolate(silentDataRequest)
        }
    }

    func finishDartBackgroundExecution(){
        if (silentDataQueue.count == 0) {
            Logger.i(TAG, "All silent actions fetched.")
            self.closeBackgroundIsolate()
        }
        else {
            Logger.i(TAG, "Remaining " + String(silentDataQueue.count) + " silents to finish")
            self.dischargeNextSilentExecution()
        }
    }
    
    func executeDartCallbackInBackgroundIsolate(_ silentDataRequest:SilentActionRequest){
        
        if self.backgroundEngine == nil {
            Logger.i(TAG, "A background message could not be handle since" +
                    "dart callback handler has not been registered")
        }
        
        var silentData:[String : Any?] = silentDataRequest.actionReceived.toMap()
        silentData[Definitions.ACTION_HANDLE] = self.silentCallbackHandle
        
        backgroundChannel?.invokeMethod(
            Definitions.CHANNEL_METHOD_SILENT_CALLBACK,
            arguments: silentData,
            result: { flutterResult in
                silentDataRequest.handler(true)
                self.finishDartBackgroundExecution()
            }
        )
    }
}
#endif

File: android/src/main/java/me/carda/awesome_notifications/DartBackgroundExecutor.java
package me.carda.awesome_notifications;

import android.content.Context;
import android.content.Intent;
import android.content.res.AssetManager;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.atomic.AtomicBoolean;

import io.flutter.FlutterInjector;
import io.flutter.embedding.engine.FlutterEngine;
import io.flutter.embedding.engine.dart.DartExecutor;
import io.flutter.embedding.engine.dart.DartExecutor.DartCallback;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.view.FlutterCallbackInformation;
import me.carda.awesome_notifications.core.AwesomeNotifications;
import me.carda.awesome_notifications.core.Definitions;
import me.carda.awesome_notifications.core.background.BackgroundExecutor;
import me.carda.awesome_notifications.core.builders.NotificationBuilder;
import me.carda.awesome_notifications.core.exceptions.AwesomeNotificationsException;
import me.carda.awesome_notifications.core.exceptions.ExceptionCode;
import me.carda.awesome_notifications.core.exceptions.ExceptionFactory;
import me.carda.awesome_notifications.core.logs.Logger;
import me.carda.awesome_notifications.core.managers.LifeCycleManager;
import me.carda.awesome_notifications.core.models.returnedData.ActionReceived;

/**
 * An background execution abstraction which handles initializing a background isolate running a
 * callback dispatcher, used to invoke Dart callbacks while backgrounded.
 */
public class DartBackgroundExecutor extends BackgroundExecutor implements MethodCallHandler {
    private static final String TAG = "DartBackgroundExec";

    private static final BlockingQueue<Intent> silentDataQueue = new LinkedBlockingDeque<Intent>();

    public static Context applicationContext;

    private AtomicBoolean isRunning = null;

    private MethodChannel backgroundChannel;
    private FlutterEngine backgroundFlutterEngine;

    public boolean isDone(){
        return isRunning != null && !isRunning.get();
    }

    public boolean runBackgroundAction(
            Context context,
            Intent silentIntent
    ){
        if (dartCallbackHandle == 0) return false;
        applicationContext = context;

        addSilentIntent(silentIntent);

        if (isRunning == null){
            isRunning = new AtomicBoolean(true);
            runBackgroundThread(dartCallbackHandle);
        }

        return true;
    }

    private static io.flutter.plugin.common.PluginRegistry.PluginRegistrantCallback
            pluginRegistrantCallback;

    /**
     * Sets the {@code io.flutter.plugin.common.PluginRegistry.PluginRegistrantCallback} used to
     * register plugins with the newly spawned isolate.
     *
     * <p>Note: this is only necessary for applications using the V1 engine embedding API as plugins
     * are automatically registered via reflection in the V2 engine embedding API. If not set,
     * background message callbacks will not be able to utilize functionality from other plugins.
     */
    public static void setPluginRegistrant(
            io.flutter.plugin.common.PluginRegistry.PluginRegistrantCallback callback) {
        pluginRegistrantCallback = callback;
    }

    private static void addSilentIntent(Intent intent){
        silentDataQueue.add(intent);
    }

    @Override
    public void onMethodCall(MethodCall call, @NonNull Result result) {
        String method = call.method;
        try {
            if (method.equals(Definitions.CHANNEL_METHOD_PUSH_NEXT_DATA)) {
                dischargeNextSilentExecution();
                result.success(true);
            } else {
                result.notImplemented();
            }
        } catch (Exception e) {
            AwesomeNotificationsException awesomeException =
                ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_UNKNOWN_EXCEPTION,
                            "An unexpected exception was found in a silent background execution",
                            ExceptionCode.DETAILED_UNEXPECTED_ERROR);
            result.error(
                    awesomeException.getCode(),
                    awesomeException.getMessage(),
                    awesomeException.getDetailedCode());
        }
    }

    public void runBackgroundThread(final Long callbackHandle) {

        if (backgroundFlutterEngine != null) {
            Logger.e(TAG, "Background isolate already started.");
            return;
        }

        // giving time to debug attach (only for tests)
//        try {
//            Thread.sleep(4000);
//        } catch (InterruptedException e) {
//            e.printStackTrace();
//        }

        final Handler handler = new Handler(Looper.getMainLooper());
        Runnable dartBgRunnable =
                new Runnable() {
                    @Override
                    public void run() {

                        Logger.i(TAG, "Initializing Flutter global instance.");

                        FlutterInjector.instance().flutterLoader().startInitialization(applicationContext.getApplicationContext());
                        FlutterInjector.instance().flutterLoader().ensureInitializationCompleteAsync(
                                applicationContext.getApplicationContext(),
                                null,
                                handler,
                                new Runnable() {
                                    @Override
                                    public void run() {
                                        String appBundlePath = FlutterInjector.instance().flutterLoader().findAppBundlePath();
                                        AssetManager assets = applicationContext.getApplicationContext().getAssets();

                                        Logger.i(TAG, "Creating background FlutterEngine instance.");
                                        backgroundFlutterEngine =
                                                new FlutterEngine(applicationContext.getApplicationContext());

                                        // We need to create an instance of `FlutterEngine` before looking up the
                                        // callback. If we don't, the callback cache won't be initialized and the
                                        // lookup will fail.
                                        FlutterCallbackInformation flutterCallback =
                                                FlutterCallbackInformation.lookupCallbackInformation(callbackHandle);

                                        if (flutterCallback == null){
                                            ExceptionFactory
                                                    .getInstance()
                                                    .registerNewAwesomeException(
                                                            TAG,
                                                            ExceptionCode.CODE_MISSING_ARGUMENTS,
                                                            "The flutter background reference for dart background action is invalid",
                                                            ExceptionCode.DETAILED_REQUIRED_ARGUMENTS+".FlutterCallbackInformation");
                                        }
                                        else {
                                            DartExecutor executor = backgroundFlutterEngine.getDartExecutor();
                                            initializeReverseMethodChannel(executor);

                                            Logger.i(TAG, "Executing background FlutterEngine instance.");
                                            DartCallback dartCallback =
                                                    new DartCallback(assets, appBundlePath, flutterCallback);
                                            executor.executeDartCallback(dartCallback);
                                        }
                                    }
                                });
                    }
                };

        handler.post(dartBgRunnable);
    }

    private void initializeReverseMethodChannel(BinaryMessenger isolate) {
        backgroundChannel = new MethodChannel(isolate, Definitions.DART_REVERSE_CHANNEL);
        backgroundChannel.setMethodCallHandler(this);
    }

    public void closeBackgroundIsolate() {
        if (!isDone()) {

            isRunning.set(false);

            Handler handler = new Handler(Looper.getMainLooper());
            Runnable dartBgRunnable =
                    new Runnable() {
                        @Override
                        public void run() {
                            Logger.i(TAG, "Shutting down background FlutterEngine instance.");

                            if (backgroundFlutterEngine != null) {
                                backgroundFlutterEngine.destroy();
                                backgroundFlutterEngine = null;
                            }

                            Logger.i(TAG, "FlutterEngine instance terminated.");
                        }
                    };

            handler.post(dartBgRunnable);
        }
    }

    public void dischargeNextSilentExecution(){
        if (!silentDataQueue.isEmpty()) {
            try {
                Intent intent = silentDataQueue.take();
                executeDartCallbackInBackgroundIsolate(intent);
            } catch (Exception e) {
                ExceptionFactory
                        .getInstance()
                        .registerNewAwesomeException(
                                TAG,
                                ExceptionCode.CODE_BACKGROUND_EXECUTION_EXCEPTION,
                                ExceptionCode.DETAILED_UNEXPECTED_ERROR+".background.silentExecution",
                                e);
            }
        }
        else {
            closeBackgroundIsolate();
        }
    }

    private void finishDartBackgroundExecution(){
        if(silentDataQueue.isEmpty()) {
            if(AwesomeNotifications.debug)
                Logger.i(TAG, "All silent data fetched.");
            closeBackgroundIsolate();
        }
        else {
            if (AwesomeNotifications.debug)
                Logger.i(TAG, "Remaining " + silentDataQueue.size() + " silents to finish");
            dischargeNextSilentExecution();
        }
    }

    private final Result dartChannelResultHandle =
        new Result() {
            @Override
            public void success(Object result) {
                finishDartBackgroundExecution();
            }

            @Override
            public void error(String errorCode, String errorMessage, Object errorDetails) {
                finishDartBackgroundExecution();
            }

            @Override
            public void notImplemented() {
                finishDartBackgroundExecution();
            }
        };

    public void executeDartCallbackInBackgroundIsolate(Intent intent) throws AwesomeNotificationsException {

        if (backgroundFlutterEngine == null) {
            Logger.i( TAG,"A background message could not be handled since " +
                    "dart callback handler has not been registered.");
            return;
        }

        ActionReceived actionReceived =
                NotificationBuilder
                        .getNewBuilder()
                        .buildNotificationActionFromIntent(
                                applicationContext,
                                intent,
                                LifeCycleManager.getApplicationLifeCycle());

        // If this intent contains a valid awesome action
        if(actionReceived != null){

            final Map<String, Object> actionData = actionReceived.toMap();
            actionData.put(
                    Definitions.ACTION_HANDLE,
                    silentCallbackHandle);

            // Handle the message event in Dart.
            backgroundChannel.invokeMethod(
                Definitions.CHANNEL_METHOD_SILENT_CALLBACK,
                actionData,
                dartChannelResultHandle);

        } else {
            Logger.e(TAG, "Silent data model not found inside Intent content.");
            finishDartBackgroundExecution();
        }
    }
}